device/transport/amqp/src/amqp.ts

```diff
@@ -74,54 +83,331 @@ export class Amqp extends EventEmitter implements Client.Transport, StableConnec
+              if (err) {
+                debug('error creating a C2D link: ' + err.toString());
+                handleResult('AMQP Transport: Could not attach link', callback)(err);
+              } else {
+                debug('C2D link created and attached successfully');
+                this._c2dLink = receiverLink;
+                this._c2dLink.on('error', this._c2dErrorListener);
+                this._c2dLink.on('message', this._c2dMessageListener);
+                callback();
+              }
+            });
+          },
+          detachC2DLink: (callback) => {
+            this._stopC2DListener(undefined, callback);
+          },
+          onDeviceMethod: (methodName, methodCallback) => {
+            /*Codes_SRS_NODE_DEVICE_AMQP_16_022: [The `onDeviceMethod` method shall call the `onDeviceMethod` method on the `AmqpDeviceMethodClient` object with the same arguments.]*/
+            this._deviceMethodClient.onDeviceMethod(methodName, methodCallback);
+            this._deviceMethodClient.attach((err) => {
+              if (err) {
+                debug('error while attaching the device method client: ' + err.toString());
+                // no need to emit anything: the general error handler registered in the constructor will do that for us.
+              }
+            });
+          }
+        },
+        disconnecting: {
+          _onEnter: (disconnectCallback, err) => {
+            let finalError = err;
+            async.series([
+              (callback) => {
+                if (this._d2cLink) {
+                  let tmpD2CLink = this._d2cLink;
+                  this._d2cLink = undefined;
+
+                  if (err) {
+                  /*Codes_SRS_NODE_DEVICE_AMQP_16_023: [The `disconnect` method shall forcefully detach all attached links if a connection error is the causing the transport to be disconnected.]*/
+                    tmpD2CLink.forceDetach(err);
+                    tmpD2CLink.removeListener('error', this._d2cErrorListener);
+                  }
+                  tmpD2CLink.detach((detachErr) => {
+                    if (detachErr) {
+                      debug('error detaching the D2C link: ' + detachErr.toString());
+                    } else {
+                      debug('D2C link detached');
+                    }
+                    tmpD2CLink.removeListener('error', this._d2cErrorListener);
+                    if (!finalError && detachErr) {
+                      finalError = translateError('error while detaching the D2C link when disconnecting', detachErr);
+                    }
+                    callback(finalError);
+                  });
+                } else {
+                  callback();
+                }
+              },
+              (callback) => {
+                if (this._c2dLink) {
+                  /*Codes_SRS_NODE_DEVICE_AMQP_16_022: [The `disconnect` method shall detach all attached links.]*/
+                  this._stopC2DListener(err, callback);
```

Comment:
>this._stopC2DListener(err, callback); [](start = 18, length = 37)

If I understand async.series, then this will abort all the following actions on failure. right?  This means that a failure stopping the C2D listener will prevent this._amqp._disconnct from being called.  Is that what we want?

Reply:
nope. bug. fixed.

---
In reply to: [141135707](https://github.com/Azure/azure-iot-sdk-node/pull/89#discussion_r141135707) [](ancestors = 141135707)